<a href="https://colab.research.google.com/github/lcbjrrr/algojust/blob/main/Pmpt04_hist_algo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algorithmic and Historical bias


- **Algorithmic bias**: Bias that arises from the design, implementation, or use of algorithms. This can be due to skewed training data or poor algorithm design and can lead to unfair outcomes in areas like recruiting or criminal justice.
Omitted variable bias: Occurs when important variables are left out of a model or analysis, leading to inaccurate results. For example, analyzing car data without considering mileage or age can lead to inaccurate conclusions about vehicle value

- **Historical bias**: Occurs when past socio-cultural prejudices and beliefs are reflected in the data and subsequent analysis. This is particularly challenging when historical data is used to train machine learning models, as the models can perpetuate and amplify these biases. An example is an AI hiring tool that learned from historically biased hiring data and ended up favoring male candidates


*The data used herein is for illustrative purposes only and does not reflect actual real-world data*.

#### Here is a trained [model](https://colab.research.google.com/github/lcbjrrr/quantai/blob/main/M2_Py_ML_RegLin.ipynb), please use it to perform preditions

## AI/ML Models

Machine Learning (ML) models are a technology where computers learn from data to find patterns and make predictions without needing explicit programming for each task. These models improve their accuracy by processing more data and adjusting how they analyze information over time.



> Read the file 'car_insurance - bias.csv' and show the dataframe



In [3]:
import pandas as pd
df = pd.read_csv('car_insurance - bias.csv')
df.head()

age  female  premium  vehicle_age  claim_last2y          v
0   23       1      580            3             0      Sedan
1   24       0      520            5             1        SUV
2   31       1      510            2             0  Hatchback
3   33       0      450            7             1      Sedan
4   40       1      440            4             0        SUV

In [2]:
#Here are the predictions. **What kind of bias do you see on those**?
#test['predictions']=model.predict(test[['age','female','vehicle_age','claim_last2y','v_Hatchback','v_SUV','v_Sedan','v_Truck']])
#test

## Training the model

The get_dummies function in pandas is used for one-hot encoding categorical variables. It converts categorical data into a format that can be easily understood by machine learning algorithms, where each category becomes a new column with binary values (0 or 1).



> Transform the categorical variable into numerical ones



In [4]:
df = pd.get_dummies(df, columns=['v'], drop_first=False)
df.head()

age  female  premium  vehicle_age  claim_last2y  v_Hatchback  v_SUV  \
0   23       1      580            3             0        False  False   
1   24       0      520            5             1        False   True   
2   31       1      510            2             0         True  False   
3   33       0      450            7             1        False  False   
4   40       1      440            4             0        False   True   

   v_Sedan  v_Truck  
0     True    False  
1    False    False  
2    False    False  
3     True    False  
4    False    False

To train a AI/ML model, in the example here a linear regression model using a training dataset, save the trained model to a file, load it back, and then use it to make predictions. The process involves importing libraries, fitting the model to data, evaluating its performance, and applying it to new data points to see the predicted outcomes. It is possible to measure the correctness in traning with the score() method, in the case here (RegLing) the r2.



> Train a linear regression model, so it learn how to predict Premium based on age, female, vehicle_age, claim_last2y, vehicle type. Then show the score (R2) of the training.



In [5]:
from sklearn.linear_model import LinearRegression

X = df[['age', 'female', 'vehicle_age', 'claim_last2y', 'v_Hatchback', 'v_SUV', 'v_Sedan', 'v_Truck']]
y = df['premium']

linreg = LinearRegression()
linreg.fit(X, y)

r2_score = linreg.score(X, y)
print(f"R2 Score of the training: {r2_score}")

R2 Score of the training: 0.9824782207152526




> Create a test dataset with 3 persons with a hatchback,
one being a man and the other two women,
The first two have a claim in the last two years,
and the other has no. All three with a 2-year-old car.
Add another 33-year-old woman, but with a 4-year-old car.
Then add two people of 66, one man and the other a woman,
both with a 4-year-old SUV, and with no claim in the last 2 years


In [6]:
import pandas as pd
test = pd.DataFrame({'age':[33,33,33,33,66,66],
                     'female':[0,1,1,1,1,0],
                     'vehicle_age':[2,2,2,4,4,4],
                     'claim_last2y':[1,1,0,0,0,0],
                     'v_Hatchback':[1,1,1,1,0,0],
                     'v_SUV':[0,0,0,0,1,1],
                     'v_Sedan':[0,0,0,0,0,0],
                     'v_Truck':[0,0,0,0,0,0] })
test

age  female  vehicle_age  claim_last2y  v_Hatchback  v_SUV  v_Sedan  \
0   33       0            2             1            1      0        0   
1   33       1            2             1            1      0        0   
2   33       1            2             0            1      0        0   
3   33       1            4             0            1      0        0   
4   66       1            4             0            0      1        0   
5   66       0            4             0            0      1        0   

   v_Truck  
0        0  
1        0  
2        0  
3        0  
4        0  
5        0



> Create predictions on this dataset and save it as a column on the test dataset.



In [7]:
test['predictions'] = linreg.predict(test[['age','female','vehicle_age','claim_last2y','v_Hatchback','v_SUV','v_Sedan','v_Truck']])
display(test)

age  female  vehicle_age  claim_last2y  v_Hatchback  v_SUV  v_Sedan  \
0   33       0            2             1            1      0        0   
1   33       1            2             1            1      0        0   
2   33       1            2             0            1      0        0   
3   33       1            4             0            1      0        0   
4   66       1            4             0            0      1        0   
5   66       0            4             0            0      1        0   

   v_Truck  predictions  
0        0   455.588702  
1        0   494.878217  
2        0   495.767483  
3        0   495.139011  
4        0   256.802817  
5        0   217.513302

*From the test dataset clearly indicate that for individuals with otherwise identical risk profiles (age, vehicle age, claim history, vehicle type), female individuals are consistently predicted to pay higher insurance premiums than male individuals. For example: 33-year-olds with a 2-year-old hatchback and a recent claim: A male is predicted to pay 455.59, while a female is predicted to pay 494.88. 66-year-olds with a 4-year-old SUV and no recent claim: A male is predicted to pay 217.51, while a female is predicted to pay 256.80.
This pattern suggests that the model is discriminating based on gender, which is a form of bias. This could stem from historical biases present in the training data (if historical premiums were higher for women without justified actuarial reasons) or from an algorithmic design that inadvertently amplifies gender as a predictive factor even when it shouldn't be*

## Linear Regression 101

In [ ]:
import pandas as pd
treino = pd.read_csv('https://raw.githubusercontent.com/lcbjrrr/algojust/refs/heads/main/grades%20-%20all.csv')
treino.head(3)

Ex1  Ex2  Ex3
0  100  100   90
1   90  100   90
2   95  100  100

In [ ]:
from sklearn.linear_model import LinearRegression
reglin = LinearRegression()
reglin.fit(treino[['Ex1','Ex2']] , treino['Ex3'])
reglin.score(treino[['Ex1','Ex2']] , treino['Ex3'])

0.9818896713333345

In [ ]:
test = pd.DataFrame({'Ex1':[80,20,35],
                     'Ex2':[80,40,40],
                     'Ex3':[35,80,90]})
test

Ex1  Ex2  Ex3
0   80   80   35
1   20   40   80
2   35   40   90

In [ ]:
preds = reglin.predict(test[['Ex1','Ex2']])
preds

array([73.12693817, 39.75624675, 34.2650204 ])

## Activity: Algorithmic and Historical bias


**Problem**

Even foundational models like linear regression can perpetuate and amplify biases present in their training data, leading to unfair or inaccurate predictions. Your task is to select a dataset from Kaggle suitable for linear regression. Train a linear regression model on your chosen dataset and generate predictions. Then analyze your predictions for evidence of bias (in case there are).

**Conclusions**

Summarize the specific biases identified in your model's predictions, detailing their impact. Discuss the effectiveness of any mitigation strategies you attempted or propose and reflect on the ethical implications of your findings. Your conclusion should articulate the challenges of achieving fairness in machine learning and suggest future research directions or practical steps for developing more equitable predictive models. In case there are no signs of bias, provide concrete evidence of it and comment on this finding.

## Appendix



```
!wget https://github.com/lcbjrrr/algojust/raw/refs/heads/main/lingreg.pkl
```





```
import joblib
model = joblib.load('lingreg.pkl')
model.coef_
```



The joblib.load function in Python quickly loads saved data like machine learning models, especially when working with large arrays. .pkl files are Python-specific files created with the pickle module to store objects in a serialized format for later use.​

With the trained model, we can perform predictions. Predictions are estimates or decisions made by trained models when given new, unseen data, using patterns learned from historical information.​





```
joblib.dump(linreg, 'lingreg.pkl')
```

